File: day7.py
import re

bag_re = re.compile(r'^(\w+ \w+) bags contain (.*)$')
bag_inner_re = re.compile(r'^[\s]?(\d) (\w+ \w+)')


def process_iter(bag_map, bag_list):
    if 'shiny gold' in bag_list:
        return 1
    
    elif not bag_list:
        return 0

    else:
        for bag in bag_list:
            found = process_iter(bag_map, bag_map[bag])
            if found == 1:
                return 1
    
    return 0



def process(bag_map):
    d = dict()
    for key, val in bag_map.items():
        d[key] = process_iter(bag_map, val)
    
    return d


def process2_iter(bag_map, agg, counter):
    if not agg:
        return counter

    else:
        for bag in agg:
            bags = bag_map[bag]
            counter = process2_iter(bag_map, bags, counter+1)
    
    return counter


def process2(bag_map):
    counter = process2_iter(bag_map, bag_map['shiny gold'], 0)

    return counter

def main():
    bag_map = dict()
    with open("./day7.in", "r") as file:
        for line in file:
            first = bag_re.match(line)
            bag = first.group(1)

            second = first.group(2).split(',')
            contained = list()
            for l in second:
                l = l.strip()
                sub = bag_inner_re.match(l)
                #print(sub)
                if sub:
                    contained.extend([sub.group(2)]*int(sub.group(1)))
            
            bag_map[bag] = contained

    import time
    t1 = time.time()
    count = 0
    r = process(bag_map)
    for key, val in r.items():
        if val == 1:
            count += 1
    t2 = time.time()
    print(count)
    print(t2-t1)

    #r = process2(bag_map)
    #print(r)

if __name__ == '__main__':
    main()

File: day7.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <unordered_map>

#include <boost/regex.hpp>
#include <chrono>


struct Bag {
    typedef std::pair<Bag, int> bagCount;

    std::string name;
    std::vector<bagCount> contains;

    Bag()
    : name{},
      contains{}
    {}

    Bag(std::string t_name, std::vector<bagCount> t_bags)
    : name{t_name},
      contains{t_bags}
      {}

    Bag(std::string t_name)
    : name{t_name},
      contains{}
      {}

};

typedef std::pair<Bag, int> BagCount;

boost::regex re_bag{"^(\\w+ \\w+) bags contain (.*)$"};
boost::regex re_contains{"^[\\s]?(\\d+) (\\w+ \\w+) bag[s,.]*?$"};

std::istream& operator>>(std::istream& is, Bag& b) {
    std::string line;
    std::getline(is, line, '\n');
    
    boost::smatch what;
    if (boost::regex_search(line, what, re_bag)) {
        b.name = what[1];
        
        std::string tmp{what[2]};
        std::vector<BagCount> contents;
        size_t pos = 0;
        while ((pos = tmp.find(',', 0)) != std::string::npos) { 
            std::string c{tmp.substr(0, pos)};
            if (boost::regex_search(c, what, re_contains)) {
                contents.push_back(
                    BagCount{
                        Bag{std::string{what[2]}}, 
                        atoi(std::string{what[1]}.c_str())}
                        );
            }
            tmp.erase(0, pos+1);
        } 
        if (boost::regex_search(tmp, what, re_contains)) {
            contents.push_back(
                BagCount{
                    Bag{std::string{what[2]}},
                    atoi(std::string{what[1]}.c_str())}
                    );
        }
        b.contains = contents;
    }
    return is;
}

typedef std::unordered_map<std::string, std::vector<BagCount>> BagLookup;

int process_iter(const BagLookup& t_lookup, const std::vector<BagCount>& t_contents) {
    if (t_contents.empty()) {
        return 0;
    }

    for (const auto& c : t_contents) {
        if (c.first.name == "shiny gold") {
            return 1;
        }
    }

    for (const auto& c: t_contents) {
        int found = process_iter(t_lookup, t_lookup.at(c.first.name));
        if (found > 0) {
            return 1;
        }
    }
    
    return 0;
}

int process(BagLookup t_lookup) {
    int cnt = 0;
    for (const auto& b : t_lookup) {
        if (process_iter(t_lookup, b.second) > 0) {
            cnt++;
        }
    }


    return cnt;
};

int main() {
    std::istream_iterator<Bag> in(std::cin);
    std::vector<Bag> bags(in, std::istream_iterator<Bag>());

    BagLookup bagLookup{};
    for (const auto& b : bags) {
        std::pair<std::string, std::vector<BagCount>> p{b.name, b.contains};
        bagLookup.insert(p);
    }

    /*
    for (const auto& b : bags) {
        std::cout << b.name << ": " ;
        for (const auto& c : b.contains) {
            std::cout << "(" << c.second << ") " << c.first.name << ", ";
        }
        std::cout << std::endl;
    }
    */

    auto t1 = std::chrono::high_resolution_clock::now();
    int cnt{process(bagLookup)};
    auto t2 = std::chrono::high_resolution_clock::now();
    
    auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(t2-t1);
    std::cout << cnt << std::endl;
    std::cout << "@" << duration.count() << " ms" << std::endl;

    return EXIT_SUCCESS;
}
